Tracker improvements + PS3EYEDriver update

rovarma opened the pull request:
This PR is kind of a mish mash of different changes:

- I've been working on integrating some useful commits from @cboulay's and @brendanwalker's psmoveapi forks. Their forks are different enough that integrating changes directly is pretty difficult (I tried...), so instead I just redid them on a clean fork. Credit goes to them. Most notable change is that the tracker can now be initialized with a settings object
- I've made some improvements of my own
- I've spent some time optimizing the PS3EYEDriver. Details about that are here: https://github.com/thp/PS3EYEDriver/pull/15 but note that the last 3 commits on this PR are required to be able to accept https://github.com/thp/PS3EYEDriver/pull/15. Without those commits, PSMoveAPI won't build with the new version of PS3EYEDriver.

Let me know if you have questions or want something done differently.

cboulay commented:
If inspirit merges https://github.com/inspirit/PS3EYEDriver/pull/21 then it should be possible to update the psmoveapi's PS3EYEDriver submodule reference to point to inspirit's repo and then thp's PS3EYEDriver is no longer necessary.

rovarma commented:
Yes, absolutely, though I don't know how active inspirit is with PRs (especially of this size), so I figured I'd also put the PR on this side.

thp commented:
Can you update the PS3EYEDriver submodule reference to point to inspirit's repo?

thp commented:
In general I'm happy with those changes, but need a day or two to have time to review the PR properly. If I don't get back to you until Friday, please ping me and I'll have a look.

rovarma commented:
Thanks, I'll update the submodule when I get home tonight.

rovarma commented:
Updated the PS3EYEDriver submodule reference.

thp commented on src/tracker/camera_control.c:
```diff
@@ -130,52 +139,76 @@ camera_control_new(int cameraID)
+#if defined(CAMERA_CONTROL_USE_CL_DRIVER)
+	return CLEyeGetCameraCount();
+#elif defined(CAMERA_CONTROL_USE_PS3EYE_DRIVER)
+	ps3eye_init();
```
Is it okay to call `ps3eye_init()` multiple times? Do we need to call `ps3eye_uninit()` and then `ps3eye_init()` for it to find newly-attached cameras? `ps3eye_uninit()` will probably fail if there are handles open, but at least we could try to re-init the library to find new cameras?

rovarma replied:
Yes, it's okay to call it multiple times (subsequent calls will be ignored). It makes sense that a client should be able to call `camera_control_count_connected` before doing anything like creating cameras, so I think we do need to `ps3eye_init` here.

This function is not currently intended to discover new cameras and I think it would be a bad idea to try to re-init the library. 

If we want to get camera discovery in there as they are plugged in (a useful feature, I think), I think it would be better to support that in PS3EYEDriver itself (somehow).

rovarma replied:
Actually, thinking about it some more, it might be a good idea to move `ps3eye_init` out of these functions altogether and perhaps move them to a (new) `psmovetracker_init` function? Makes it more explicit and less lazy. 

On the other hand, it would be a pretty big API change...

thp commented on src/tracker/camera_control.c:
```diff
@@ -130,52 +139,76 @@ camera_control_new(int cameraID)
+	return ps3eye_count_connected();
+#else
+	// Don't know how to get number of connected cameras through opencv...
+	return 0;
```
This suggest a way to do it with OpenCV:

http://answers.opencv.org/question/6696/videocapture-count-connected-cameras-windows-vs-os-x/

rovarma replied:
Thanks, I'll make this change tonight.

thp commented on src/tracker/psmove_tracker.c:
```diff
@@ -56,41 +56,51 @@
 
-/* Only re-use color mappings "younger" than 2 hours */
-#define COLOR_MAPPING_MAX_AGE (2*60*60)
+static const PSMoveTrackerSettings tracker_default_settings = {
```
This looks much nicer indeed than all the `#define`s above :)

rovarma replied:
Thanks, but I can't take credit for it, this was either @cboulay or @brendanwalker :)

thp commented on src/tracker/psmove_tracker.c:
```diff
@@ -401,7 +379,21 @@ psmove_tracker_remember_color(PSMoveTracker *tracker, struct PSMove_RGBValue rgb
+void
+psmove_tracker_settings_set_default(PSMoveTrackerSettings *settings)
+{
+    *settings = tracker_default_settings;  // Doesn't work.
```
Doesn't work? Should work?

rovarma replied:
I..err...actually, don't know. I'm fairly sure it does work, since I'm using it, so the comment is probably wrong. Will have a look tonight. Maybe @cboulay has some idea?

thp commented:
Added some comments - please have a look. Also, please replace calls to `printf()` with a corresponding call to `psmove_DEBUG()`, `psmove_WARNING()` or something similar (see [here](https://github.com/thp/psmoveapi/blob/master/src/psmove_private.h#L58)).

rovarma commented:
Thanks for the comments. I'll try to get to them tonight.

rovarma commented:
I removed the comment in `psmove_tracker_settings_set_default` (it does work). I also removed the `printf` statement as it was a bit useless anyway. Are there any other `printf` statements that should be removed?

I tried making the change to implement `camera_control_count_connected` for OpenCV but didn't finish it. The suggested usage of `cv::VideoCapture` to count the number of cameras is a C++ only class (as far I as can tell, there is no C equivalent) and according to [StackOverflow](http://stackoverflow.com/questions/7322939/how-to-count-cameras-in-opencv-2-3) sort of dubious anyway:

>OpenCV still has no API to enumerate the cameras or get the number of available devices. See this ticket on OpenCV bug tracker for details.

>Behavior of VideoCapture is undefined for device numbers greater then number of devices connected and depends from API used to communicate with your camera. See OpenCV 2.3 (C++,QtGui), Problem Initializing some specific USB Devices and Setups for the list of APIs used in OpenCV.

Given this, I think I'd prefer it returning 0, instead of putting in untested code in there that may not work anyway. What do you think?

thp commented:
Yes, returning 0 for OpenCV is good in that case (or -1, so we can distinguish between "no cameras connected" and "don't know how many cameras are connected").

thp commented:
I'm actually going to merge this now and then fix the rest myself, hope that's okay for you :)

rovarma commented:
Sure, that's fine. Thanks!
